core/ArchiveProcessor/RecordBuilder.php

```diff
@@ -0,0 +1,255 @@
+
+            $archiveProcessor->aggregateDataTableRecords(
+                $record->getName(),
+                $maxRowsInTable,
+                $maxRowsInSubtable,
+                $columnToSortByBeforeTruncation,
+                $this->columnAggregationOps
+            );
+        }
+
+        if (!empty($numericRecords)) {
+            $numericMetrics = array_map(function (Record $r) { return $r->getName(); }, $numericRecords);
+            if (!empty($requestedReports)) {
+                $numericMetrics = array_filter($numericMetrics, function ($name) use ($requestedReports, $foundRequestedReports) {
+                    return in_array($name, $requestedReports) && !in_array($name, $foundRequestedReports);
+                });
+            }
+            $archiveProcessor->aggregateNumericMetrics($numericMetrics, $this->columnAggregationOps);
+        }
+    }
+
+    /**
+     * Returns metadata for records primarily used when aggregating over non-day periods. Every numeric/blob
+     * record your RecordBuilder creates should have an associated piece of record metadata.
+     *
+     * @return Record[]
+     */
+    public abstract function getRecordMetadata(ArchiveProcessor $archiveProcessor): array;
+
+    /**
+     * Derived classes should define this method to aggregate log data for a single day and return the records
+     * to store indexed by record names.
+     *
+     * @return (DataTable|int|float|string)[] Record values indexed by their record name, eg, `['MyPlugin_MyRecord' => new DataTable()]`
+     */
+    protected abstract function aggregate(ArchiveProcessor $archiveProcessor): array;
+
+    private function insertRecord(ArchiveProcessor $archiveProcessor, $recordName, DataTable\DataTableInterface $record,
+                                  ?int $maxRowsInTable, ?int $maxRowsInSubtable, ?string $columnToSortByBeforeTruncation): void
+    {
+        $serialized = $record->getSerialized(
+            $maxRowsInTable ?: $this->maxRowsInTable,
+            $maxRowsInSubtable ?: $this->maxRowsInSubtable,
+            $columnToSortByBeforeTruncation ?: $this->columnToSortByBeforeTruncation
+        );
+        $archiveProcessor->insertBlobRecord($recordName, $serialized);
+        unset($serialized);
+    }
+
+    public function getMaxRowsInTable(): ?int
+    {
+        return $this->maxRowsInTable;
+    }
+
+    public function getMaxRowsInSubtable(): ?int
+    {
+        return $this->maxRowsInSubtable;
+    }
+
+    public function getColumnToSortByBeforeTruncation(): ?string
+    {
+        return $this->columnToSortByBeforeTruncation;
+    }
+
+    public function getPluginName(): ?string
+    {
+        // TODO: consider extracting to a reusable method or a trait, or use another approach to getting plugin's name
+        $className = get_class($this);
+        $parts = explode('\\', $className);
+        $plugin = $parts[2];
+        return $plugin;
+    }
```

Comment:
Can that method actually return null? I guess that could only happen if the class of the record builder would not be in the correct namespace of a plugin. But in that case it shouldn't be used at all. And the method would actually trigger a warning as `$parts[2]` might then be undefined.
```suggestion
    public function getPluginName(): string
    {
        // TODO: consider extracting to a reusable method or a trait, or use another approach to getting plugin's name
        $className = get_class($this);
        $parts = explode('\\', $className);
        $plugin = $parts[2] ?? '';
        return $plugin;
    }
```